### PR TITLE
Optimize CIOS for moduli with spare bits

### PR DESCRIPTION
**Pull request by schouhy**
# Optimize CIOS for moduli with spare bits

## Description

Solves Issue #506

Screenshot 2023-07-31 at 16 26 33

## Type of change

- [ ] New feature
- [ ] Bug fix
- [x] Optimization

## Checklist
- [x] This change is an Optimization
  - [x] Benchmarks added/run

### CODE REVIEW

File: math/src/unsigned_integer/montgomery.rs [hunk 1]
```diff
@@ -75,6 +75,71 @@ impl MontgomeryAlgorithms {
...
+            while j > 0 {
+                j -= 1;
+                cs = t[j] as u128 + m * (q.limbs[j] as u128) + c;
+                c = cs >> 64;
+                t[j + 1] = ((cs << 64) >> 64) as u64;
+            }
+
+            // (C,t[N-1]) := t[N] + C
+            cs = (t_extra as u128) + c;
+            t[0] = ((cs << 64) >> 64) as u64;
```
Comment (Oppen) [inline math/src/unsigned_integer/montgomery.rs:133]: Is this correct? Shouldn't `cs` be stored at the most significant bit of `t[0]`?

File: math/src/unsigned_integer/montgomery.rs [hunk 2]
```diff
@@ -144,6 +209,32 @@ impl MontgomeryAlgorithms {
...
+            let m = U384::from_hex_unchecked("cdb061954fdd36e5176f50dbdcfd349570a29ce1"); // this is prime
+            let mu: u64 = 16085280245840369887; // negative of the inverse of `m` modulo 2^{64}
+            assert_eq!(
+                MontgomeryAlgorithms::cios(&x, &y, &m, &mu),
+                MontgomeryAlgorithms::cios_optimized_for_moduli_with_one_spare_bit(&x, &y, &m, &mu)
+            );
+        }
+
+        #[test]
+        fn cios_vs_cios_square(a in any::<[u64; 6]>()) {
```
Comment (Oppen) [inline math/src/unsigned_integer/montgomery.rs:228]: Nit:
```suggestion
        fn cios_vs_sos_square(a in any::<[u64; 6]>()) {
```

Comment (Oppen) [inline math/src/unsigned_integer/montgomery.rs:228]: Nit:
```suggestion
        fn cios_vs_sos_square(a in any::<[u64; 6]>()) {
```

File: math/src/unsigned_integer/montgomery.rs (hunk 1 above)
Comment (schouhy) [reply math/src/unsigned_integer/montgomery.rs:133]: The comments in the code follow the paper convention where `t[N]` is the most significant limb. In our case that is `t[0]`. This is super confusing. I'll change it.

**Review by MauroToscano — APPROVED** [review]
Fuzzers are running properly. No mistakes are detected

**Pull request merged**